src/adapter/z-stack/adapter/adapter-backup.ts

```diff
@@ -0,0 +1,430 @@
+            }
+        }
+
+        /* recover coordinator ieee address if security entries were recovered */
+        if (tclkTable.usedCount > 0 || apsLinkKeyDataTable.usedCount > 0) {
+            const reversedAdapterIeee = Buffer.from(backup.coordinatorIeeeAddress).reverse();
+            await this.nv.writeItem(NvItemsIds.EXTADDR, reversedAdapterIeee);
+        }
+
+        /* write update nib */
+        await this.nv.writeItem(NvItemsIds.NIB, nib);
+
+        /* write network key info */
+        await this.nv.updateItem(NvItemsIds.NWK_ACTIVE_KEY_INFO, keyDescriptor.serialize());
+        await this.nv.updateItem(NvItemsIds.NWK_ALTERN_KEY_INFO, keyDescriptor.serialize());
+
+        /* write tclk seed if present */
+        if (backup.znp?.trustCenterLinkKeySeed) {
+            await this.nv.writeItem(NvItemsIds.TCLK_SEED, backup.znp.trustCenterLinkKeySeed);
+        }
+
+        /* write network security material table (nwk frame counters) */
+        if (version === ZnpVersion.zStack3x0) {
+            await this.nv.writeTable("extended", NvSystemIds.ZSTACK, NvItemsIds.EX_NWK_SEC_MATERIAL_TABLE, nwkSecurityMaterialTable);
+        } else {
+            await this.nv.writeTable("legacy", NvItemsIds.LEGACY_NWK_SEC_MATERIAL_TABLE_START, nwkSecurityMaterialTable);
+        }
+
+        /* write address manager table */
+        if (version === ZnpVersion.zStack3x0) {
+            await this.nv.writeTable("extended", NvSystemIds.ZSTACK, NvItemsIds.ZCD_NV_EX_ADDRMGR, addressManagerTable);
+        } else {
+            await this.nv.writeItem(NvItemsIds.ADDRMGR, addressManagerTable);
+        }
+
+        /* write security manager table */
+        await this.nv.writeItem(NvItemsIds.APS_LINK_KEY_TABLE, securityManagerTable);
+
+        /* write aps link key data table */
+        if (version === ZnpVersion.zStack3x0) {
+            await this.nv.writeTable("extended", NvSystemIds.ZSTACK, NvItemsIds.ZCD_NV_EX_APS_KEY_DATA_TABLE, apsLinkKeyDataTable);
+        } else {
+            await this.nv.writeTable("legacy", NvItemsIds.APS_LINK_KEY_DATA_START, apsLinkKeyDataTable);
+        }
+
+        /* write tclk table */
+        if (version === ZnpVersion.zStack3x0) {
+            await this.nv.writeTable("extended", NvSystemIds.ZSTACK, NvItemsIds.EX_TCLK_TABLE, tclkTable);
+        } else {
+            await this.nv.writeTable("legacy", NvItemsIds.LEGACY_TCLK_TABLE_START, tclkTable);
+        }
+    }
+
+    /**
+     * Acquires ZNP version internal to `zigbee-herdsman` from controller.
+     * 
+     * *If Z-Stack 1.2 controller is detected an error is thrown, since Z-Stack 1.2 backup
+     * and restore procedures are not supported.*
+     */
+    private async getAdapterVersion(): Promise<ZnpVersion> {
+        const versionResponse = await this.znp.request(Subsystem.SYS, "version", {});
+        const version: ZnpVersion = versionResponse.payload.product;
+        /* istanbul ignore next */
+        if (version === ZnpVersion.zStack12) {
```

Comment:
I thought backups for 1.2 were supported after this? (since migration from 1.2 -> 3.0 should be possible after this).

Reply:
Well from what I understood there basically isn't much to backup on 1.2. You just feed in network params and the network should work. So basically migration should be possible by fresh commissioning the network on 3.0.

I may look deeper into this if you wish, however I was able to migrate 1.2 to 3.0 without having anything backed up on the 1.2 network.